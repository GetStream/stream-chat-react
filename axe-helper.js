/* eslint-disable no-undef */
const { configureAxe } = require('jest-axe');

module.exports.axe = configureAxe({
  // FIXME: this is a temporary measure, the SDK needs fixing
  rules: { 'nested-interactive': { enabled: false } },
});
