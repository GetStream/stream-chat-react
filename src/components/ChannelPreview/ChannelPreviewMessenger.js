// @ts-check
import React, { useRef } from 'react';
// eslint-disable-next-line import/no-unresolved
import PropTypes from 'prop-types';
import { truncate } from '../../utils';

import { Avatar as DefaultAvatar } from '../Avatar';

/**
 * Used as preview component for channel item in [ChannelList](#channellist) component.
 * Its best suited for messenger type chat.
 *
 * @example ../../docs/ChannelPreviewMessenger.md
 * @type {import('types').ChannelPreviewMessenger}
 */
const ChannelPreviewMessenger = (props) => {
  const { Avatar = DefaultAvatar } = props;
  /** @type {React.MutableRefObject<HTMLButtonElement | null>} Typescript syntax */
  const channelPreviewButton = useRef(null);
  const unreadClass =
    props.unread >= 1 ? 'str-chat__channel-preview-messenger--unread' : '';
  const activeClass = props.active
    ? 'str-chat__channel-preview-messenger--active'
    : '';
  const onSelectChannel = () => {
    props.setActiveChannel(props.channel, props.watchers);
    if (channelPreviewButton?.current) {
      channelPreviewButton.current.blur();
    }
  };

  return (
    <button
      onClick={onSelectChannel}
      ref={channelPreviewButton}
      className={`str-chat__channel-preview-messenger ${unreadClass} ${activeClass}`}
      data-testid="channel-preview-button"
    >
      <div className="str-chat__channel-preview-messenger--left">
        {
          <Avatar
            image={props.displayImage}
            name={props.displayTitle}
            size={40}
          />
        }
      </div>
      <div className="str-chat__channel-preview-messenger--right">
        <div className="str-chat__channel-preview-messenger--name">
          <span>{props.displayTitle}</span>
        </div>
        <div className="str-chat__channel-preview-messenger--last-message">
          {truncate(props.latestMessage, props.latestMessageLength)}
        </div>
      </div>
    </button>
  );
};

ChannelPreviewMessenger.propTypes = {
  /** **Available from [chat context](https://getstream.github.io/stream-chat-react/#chat)** */
  channel: PropTypes.object.isRequired,
  /** Current selected channel object */
  activeChannel: PropTypes.object,
  /**
   * Custom UI component to display user avatar
   *
   * Defaults to and accepts same props as: [Avatar](https://github.com/GetStream/stream-chat-react/blob/master/src/components/Avatar/Avatar.js)
   * */
  Avatar: /** @type {PropTypes.Validator<React.ElementType<import('types').AvatarProps>>} */ (PropTypes.elementType),
  /** Setter for selected channel */
  setActiveChannel: PropTypes.func.isRequired,
  /**
   * Object containing watcher parameters
   * @see See [Pagination documentation](https://getstream.io/chat/docs/#channel_pagination) for a list of available fields for sort.
   * */
  watchers: PropTypes.object,
  /** Number of unread messages */
  unread: PropTypes.number,
  /** If channel of component is active (selected) channel */
  active: PropTypes.bool,
  /** Latest message's text. */
  latestMessage: PropTypes.string,
  /** Length of latest message to truncate at */
  latestMessageLength: PropTypes.number,
  /** Title of channel to display */
  displayTitle: PropTypes.string,
  /** Image of channel to display */
  displayImage: PropTypes.string,
};

ChannelPreviewMessenger.defaultProps = {
  latestMessageLength: 14,
};

export default React.memo(ChannelPreviewMessenger);
