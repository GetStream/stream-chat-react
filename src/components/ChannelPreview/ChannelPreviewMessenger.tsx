import React, { useRef } from 'react';
import clsx from 'clsx';

import { Avatar as DefaultAvatar } from '../Avatar';

import type { ChannelPreviewUIComponentProps } from './ChannelPreview';

import type { DefaultStreamChatGenerics } from '../../types/types';

const UnMemoizedChannelPreviewMessenger = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: ChannelPreviewUIComponentProps<StreamChatGenerics>,
) => {
  const {
    active,
    Avatar = DefaultAvatar,
    channel,
    className: customClassName = '',
    displayImage,
    displayTitle,
    latestMessage,
    onSelect: customOnSelectChannel,
    setActiveChannel,
    unread,
    watchers,
  } = props;

  const channelPreviewButton = useRef<HTMLButtonElement | null>(null);

  const avatarName =
    displayTitle || channel.state.messages[channel.state.messages.length - 1]?.user?.id;

  const onSelectChannel = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (customOnSelectChannel) {
      customOnSelectChannel(e);
    } else if (setActiveChannel) {
      setActiveChannel(channel, watchers);
    }
    if (channelPreviewButton?.current) {
      channelPreviewButton.current.blur();
    }
  };

  return (
    <button
      aria-label={`Select Channel: ${displayTitle || ''}`}
      aria-selected={active}
      className={clsx(
        `str-chat__channel-preview-messenger str-chat__channel-preview`,
        active && 'str-chat__channel-preview-messenger--active',
        unread && unread >= 1 && 'str-chat__channel-preview-messenger--unread',
        customClassName,
      )}
      data-testid='channel-preview-button'
      onClick={onSelectChannel}
      ref={channelPreviewButton}
      role='option'
    >
      <div className='str-chat__channel-preview-messenger--left'>
        <Avatar image={displayImage} name={avatarName} size={40} />
      </div>
      <div className='str-chat__channel-preview-messenger--right str-chat__channel-preview-end'>
        <div className='str-chat__channel-preview-end-first-row'>
          <div className='str-chat__channel-preview-messenger--name'>
            <span>{displayTitle}</span>
          </div>
          {!!unread && (
            <div className='str-chat__channel-preview-unread-badge' data-testid='unread-badge'>
              {unread}
            </div>
          )}
        </div>
        <div className='str-chat__channel-preview-messenger--last-message'>{latestMessage}</div>
      </div>
    </button>
  );
};

/**
 * Used as preview component for channel item in [ChannelList](#channellist) component.
 * Its best suited for messenger type chat.
 */
export const ChannelPreviewMessenger = React.memo(
  UnMemoizedChannelPreviewMessenger,
) as typeof UnMemoizedChannelPreviewMessenger;
