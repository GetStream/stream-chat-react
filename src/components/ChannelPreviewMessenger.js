import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { Avatar } from './Avatar';

import truncate from 'lodash/truncate';

/**
 * Used as preview component for channel item in [ChannelList](#channellist) component.
 * Its best suited for messenger type chat.
 *
 * @example ./docs/ChannelPreviewMessenger.md
 * @extends PureComponent
 */
export class ChannelPreviewMessenger extends PureComponent {
  static propTypes = {
    /** **Available from [chat context](https://getstream.github.io/stream-chat-react/#chat)** */
    setActiveChannel: PropTypes.func,
    /** **Available from [chat context](https://getstream.github.io/stream-chat-react/#chat)** */
    channel: PropTypes.object,
    closeMenu: PropTypes.func,
    unread: PropTypes.number,
    /** If channel of component is active (selected) channel */
    active: PropTypes.bool,
    latestMessage: PropTypes.string,
  };

  channelPreviewButton = React.createRef();

  onSelectChannel = () => {
    this.props.setActiveChannel(this.props.channel, this.props.watchers);
    this.channelPreviewButton.current.blur();
    this.props.closeMenu();
  };
  render() {
    const unreadClass =
      this.props.unread >= 1
        ? 'str-chat__channel-preview-messenger--unread'
        : '';
    const activeClass = this.props.active
      ? 'str-chat__channel-preview-messenger--active'
      : '';

    const { channel } = this.props;

    return (
      <button
        onClick={this.onSelectChannel}
        ref={this.channelPreviewButton}
        className={`str-chat__channel-preview-messenger ${unreadClass} ${activeClass}`}
      >
        <div className="str-chat__channel-preview-messenger--left">
          {<Avatar image={channel.data.image} size={40} />}
        </div>
        <div className="str-chat__channel-preview-messenger--right">
          <div className="str-chat__channel-preview-messenger--name">
            <span>{channel.data.name}</span>
          </div>
          <div className="str-chat__channel-preview-messenger--last-message">
            {channel.state.messages[0]
              ? truncate(this.props.latestMessage, { length: 14 })
              : 'Nothing yet...'}
          </div>
        </div>
      </button>
    );
  }
}
