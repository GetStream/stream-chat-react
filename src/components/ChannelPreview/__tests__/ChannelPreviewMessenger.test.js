import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { toHaveNoViolations } from 'jest-axe';
import { axe } from '../../../../axe-helper';
expect.extend(toHaveNoViolations);

import {
  generateChannel,
  generateUser,
  getOrCreateChannelApi,
  getTestClientWithUser,
  useMockedApis,
} from 'mock-builders';

import { ChannelPreviewMessenger } from '../ChannelPreviewMessenger';

describe('ChannelPreviewMessenger', () => {
  const clientUser = generateUser();
  let chatClient;
  let channel;
  const renderComponent = (props) => (
    <div aria-label='Select Channel' role='listbox'>
      <ChannelPreviewMessenger
        channel={channel}
        displayImage='https://randomimage.com/src.jpg'
        displayTitle='Channel name'
        latestMessage='Latest message!'
        setActiveChannel={jest.fn()}
        unread={10}
        {...props}
      />
    </div>
  );

  const initializeChannel = async (c) => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useMockedApis(chatClient, [getOrCreateChannelApi(c)]);

    channel = chatClient.channel('messaging');

    await channel.watch();
  };

  beforeEach(async () => {
    chatClient = await getTestClientWithUser(clientUser);
    await initializeChannel(generateChannel());
  });

  it('should render correctly', () => {
    const tree = renderer.create(renderComponent()).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call setActiveChannel on click', async () => {
    const setActiveChannel = jest.fn();
    const { container, getByTestId } = render(
      renderComponent({
        setActiveChannel,
        watchers: {},
      }),
    );

    await waitFor(() => {
      expect(getByTestId('channel-preview-button')).toBeInTheDocument();
    });

    fireEvent.click(getByTestId('channel-preview-button'));

    await waitFor(() => {
      // eslint-disable-next-line jest/prefer-called-with
      expect(setActiveChannel).toHaveBeenCalledTimes(1);
      expect(setActiveChannel).toHaveBeenCalledWith(channel, {});
    });
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should render custom class name', () => {
    const className = 'custom-xxx';
    const { container } = render(renderComponent({ className }));
    expect(container.querySelector(`.${className}`)).toBeInTheDocument();
  });

  it('should call custom onSelect function', () => {
    const onSelect = jest.fn();
    render(renderComponent({ onSelect }));
    const previewButton = screen.queryByTestId('channel-preview-button');
    fireEvent.click(previewButton);
    expect(onSelect).toHaveBeenCalledTimes(1);
  });
});
