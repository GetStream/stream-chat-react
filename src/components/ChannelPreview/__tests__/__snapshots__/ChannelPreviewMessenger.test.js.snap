// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelPreviewMessenger should render correctly 1`] = `
<div
  aria-label="Select Channel"
  role="listbox"
>
  <button
    aria-label="Select Channel: Channel name"
    className="str-chat__channel-preview-messenger str-chat__channel-preview str-chat__channel-preview-messenger--unread"
    data-testid="channel-preview-button"
    onClick={[Function]}
    role="option"
  >
    <div
      className="str-chat__channel-preview-messenger--left"
    >
      <div
        className="str-chat__avatar str-chat__avatar--circle str-chat__message-sender-avatar"
        data-testid="avatar"
        onClick={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "flexBasis": "40px",
            "fontSize": "20px",
            "height": "40px",
            "lineHeight": "40px",
            "width": "40px",
          }
        }
        title="Channel name"
      >
        <img
          alt="C"
          className="str-chat__avatar-image"
          data-testid="avatar-img"
          onError={[Function]}
          onLoad={[Function]}
          src="https://randomimage.com/src.jpg"
          style={
            Object {
              "flexBasis": "40px",
              "height": "40px",
              "objectFit": "cover",
              "width": "40px",
            }
          }
        />
      </div>
    </div>
    <div
      className="str-chat__channel-preview-messenger--right str-chat__channel-preview-end"
    >
      <div
        className="str-chat__channel-preview-end-first-row"
      >
        <div
          className="str-chat__channel-preview-messenger--name"
        >
          <span>
            Channel name
          </span>
        </div>
        <div
          className="str-chat__channel-preview-unread-badge"
          data-testid="unread-badge"
        >
          10
        </div>
      </div>
      <div
        className="str-chat__channel-preview-messenger--last-message"
      >
        Latest message!
      </div>
    </div>
  </button>
</div>
`;
